aiokafka/fetcher.py

```diff
@@ -0,0 +1,567 @@
+                # this can happen when a partition consumption paused before
+                # fetched records are returned
+                log.debug(
+                    "Not returning fetched records for assigned partition"
+                    " %s since it is no longer fetchable", tp)
+
+            elif fetch_offset == position:
+                log.debug(
+                    "Returning fetched records at offset %d for assigned"
+                    " partition %s", position, tp)
+                for msg in self._unpack_message_set(tp, messages):
+                    # Because we are in a generator, it is possible for
+                    # subscription state to change between yield calls
+                    # so we need to re-check on each loop
+                    # this should catch assignment changes, pauses
+                    # and resets via seek_to_beginning / seek_to_end
+                    if tp not in self._subscriptions.assignment:
+                        log.debug("Not returning fetched records"
+                                  " bcs partitions are unassigned")
+                        break
+                    position = self._subscriptions.assignment[tp].position
+                    if not self._subscriptions.is_fetchable(tp):
+                        log.debug(
+                            "Not returning fetched records for partition %s"
+                            " since it is no longer fetchable", tp)
+                        break
+
+                    if self._subscriptions.needs_partition_assignment:
+                        log.debug("Not returning fetched records"
+                                  " bcs partitions are unassigned")
+                        break
+
+                    # Compressed messagesets may include earlier messages
+                    # It is also possible that the user called seek()
+                    elif msg.offset != position:
+                        continue
+
+                    self._subscriptions.assignment[tp].position += 1
+                    yield msg
+            else:
+                # these records aren't next in line based on the last consumed
+                # position, ignore them they must be from an obsolete request
+                log.debug("Ignoring fetched records for %s at offset %s",
+                          tp, fetch_offset)
+
+    def next_record(self):
+        while True:
+            while self._iterator is None and self._records.empty():
+                return None
+
+            if not self._iterator:
+                self._iterator = self._message_generator()
+            try:
+                return next(self._iterator)
+            except StopIteration:
+                self._iterator = None
+                continue
+
+    def fetched_records(self):
+        """Returns previously fetched records and updates consumed offsets."""
+        drained = collections.defaultdict(list)
+        for message in self._message_generator():
```

Comment:
what do you mean? we parse messages in generator and "save" it in dict{topicPartition: list of messages} .. 
what do you suggest?